File: src/main/java/com/example/the_machine/operator/registry/SemanticSearchOperatorRegistry.java
package com.example.the_machine.operator.registry;

import com.example.the_machine.common.HashCalculator;
import com.example.the_machine.db.OperatorEmbedding;
import com.example.the_machine.operator.Operator;
import com.example.the_machine.operator.OperatorSpec;
import com.example.the_machine.repo.OperatorEmbeddingRepository;
import com.example.the_machine.service.EmbeddingService;
import jakarta.annotation.PostConstruct;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Component;

/**
 * Semantic search-enabled operator registry that uses vector embeddings for intelligent operator
 * discovery and matching. Discovers operators from the Spring context, generates embeddings for
 * their descriptions, and provides semantic search capabilities.
 */
@Component
@RequiredArgsConstructor
@Slf4j
public class SemanticSearchOperatorRegistry implements OperatorRegistry {

  private final OperatorEmbeddingRepository embeddingRepository;
  private final EmbeddingService embeddingService;
  private final HashCalculator hashCalculator;
  private final List<Operator> operators;

  private final Map<String, Operator> operatorsByName = new ConcurrentHashMap<>();

  @PostConstruct
  public void initialize() {
    discoverAndIndexOperators();
  }

  /**
   * Discovers operators and generates embeddings for new/updated operators.
   */
  private void discoverAndIndexOperators() {
    log.info("Generating operator embeddings...");

    final var newEmbeddings = new ArrayList<OperatorEmbedding>();

    for (final var operator : operators) {
      try {
        final var spec = operator.spec();
        final var operatorName = spec.name();

        // Single source of truth - cache only operators
        operatorsByName.put(operatorName, operator);

        // Use ToolSpec's existing description method + context
        final var descriptionText = createEmbeddingText(spec);
        final var existingEmbedding = embeddingRepository.findByOperatorName(operatorName);
        final var textHash = hashCalculator.calculateSha256Hash(descriptionText);

        if (existingEmbedding.isEmpty() || !existingEmbedding.get().getHash().equals(textHash)) {
          final var embedding = embeddingService.generateEmbedding(descriptionText);

          newEmbeddings.add(OperatorEmbedding.builder()
              .operatorName(operatorName)
              .embedding(embedding)
              .hash(textHash)
              .build());
        }

      } catch (Exception e) {
        log.error("Failed to process operator {}: {}", operator.getClass().getSimpleName(),
            e.getMessage(), e);
      }
    }

    if (!newEmbeddings.isEmpty()) {
      embeddingRepository.saveAll(newEmbeddings);
      log.info("Generated embeddings for {} operators", newEmbeddings.size());
    }

    log.info("Registry initialization complete. Indexed {} operators", operatorsByName.size());
  }

  /**
   * Creates embedding text from ToolSpec using existing methods.
   */
  private String createEmbeddingText(OperatorSpec spec) {
    final var text = new StringBuilder();
    text.append("Tool: ").append(spec.name());

    // ToolSpec already has getDescription()!
    if (spec.description() != null && !spec.description().trim().isEmpty()) {
      text.append(" - ").append(spec.description());
    }

    // Add I/O context for better matching
    if (!spec.getInputKeys().isEmpty()) {
      text.append(" | Inputs: ").append(String.join(", ", spec.getInputKeys()));
    }

    if (!spec.getOutputKeys().isEmpty()) {
      text.append(" | Outputs: ").append(String.join(", ", spec.getOutputKeys()));
    }

    return text.toString();
  }

  /**
   * Performs semantic search for operators based on task description.
   *
   * @param taskDescription the task description to search for
   * @return list of tool specs ordered by similarity
   */
  @Override
  public List<OperatorSpec> searchByTask(String taskDescription) {
    final var maxResults = 3;

    try {
      final var queryEmbedding = embeddingService.generateEmbedding(taskDescription);
      final var pageable = PageRequest.of(0, maxResults);

      final var results = embeddingRepository.findTopByEmbeddingSimilarity(queryEmbedding,
          pageable);

      return results.stream()
          .map(embedding -> {
            final var operator = operatorsByName.get(embedding.getOperatorName());
            return operator != null ? operator.spec() : null; // Single source of truth!
          })
          .filter(Objects::nonNull)
          .collect(Collectors.toList());

    } catch (Exception e) {
      log.error("Semantic search failed for query: {}", taskDescription, e);
      return new ArrayList<>();
    }
  }

  @Override
  public Optional<Operator> get(String name) {
    return Optional.ofNullable(operatorsByName.get(name));
  }

  @Override
  public int size() {
    return operatorsByName.size();
  }
}

File: src/main/kotlin/com/example/the_machine/operator/registry/SemanticSearchOperatorRegistry.kt
package com.example.the_machine.operator.registry

import com.example.the_machine.domain.OperatorEmbedding
import com.example.the_machine.operator.Operator
import com.example.the_machine.operator.ToolSpec
import com.example.the_machine.repo.OperatorEmbeddingRepository
import com.example.the_machine.service.EmbeddingService
import com.example.the_machine.service.HashCalculationService
import jakarta.annotation.PostConstruct
import mu.KotlinLogging
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Component
import java.util.*
import java.util.concurrent.ConcurrentHashMap

/**
 * Semantic search-enabled operator registry that uses vector embeddings for intelligent operator
 * discovery and matching. Discovers operators via ServiceLoader, generates embeddings for their
 * descriptions, and provides semantic search capabilities.
 */
@Component
class SemanticSearchOperatorRegistry(
  private val embeddingRepository: OperatorEmbeddingRepository,
  private val embeddingService: EmbeddingService,
  private val hashCalculationService: HashCalculationService
) : OperatorRegistry {

  companion object {

    private val log = KotlinLogging.logger {}
  }

  private val operatorsByName = ConcurrentHashMap<String, Operator>()

  @PostConstruct
  fun initialize() {
    discoverAndIndexOperators()
  }

  /**
   * Discovers operators using ServiceLoader and generates embeddings for new operators.
   */
  private fun discoverAndIndexOperators() {
    log.info("Discovering operators and generating embeddings...")

    val serviceLoader = ServiceLoader.load(Operator::class.java)
    val newEmbeddings = mutableListOf<OperatorEmbedding>()

    serviceLoader.forEach { operator ->
      try {
        val spec = operator.spec()
        val operatorName = spec.name

        // Single source of truth - cache only operators
        operatorsByName[operatorName] = operator

        // Use ToolSpec's existing description method + context
        val descriptionText = createEmbeddingText(spec)
        val existingEmbedding = embeddingRepository.findByOperatorName(operatorName)
        val textHash = hashCalculationService.calculateSha256Hash(descriptionText)

        if (existingEmbedding.isEmpty || existingEmbedding.get().hash != textHash) {
          val embedding = embeddingService.generateEmbedding(descriptionText)

          newEmbeddings.add(
            OperatorEmbedding(
              operatorName = operatorName,
              embedding = embedding,
              hash = textHash
            )
          )
        }
      } catch (e: Exception) {
        log.error("Failed to process operator {}: {}", operator.javaClass.simpleName, e.message, e)
      }
    }

    if (newEmbeddings.isNotEmpty()) {
      embeddingRepository.saveAll(newEmbeddings)
      log.info("Generated embeddings for {} operators", newEmbeddings.size)
    }

    log.info("Registry initialization complete. Indexed {} operators", operatorsByName.size)
  }

  /**
   * Creates embedding text from ToolSpec using existing methods.
   */
  private fun createEmbeddingText(spec: ToolSpec): String {
    val text = StringBuilder().apply {
      append("Tool: ").append(spec.name)

      // ToolSpec already has getDescription()!
      spec.description?.takeIf { it.isNotBlank() }?.let { description ->
        append(" - ").append(description)
      }

      // Add I/O context for better matching
      if (spec.getInputKeys().isNotEmpty()) {
        append(" | Inputs: ").append(spec.getInputKeys().joinToString(", "))
      }

      if (spec.getOutputKeys().isNotEmpty()) {
        append(" | Outputs: ").append(spec.getOutputKeys().joinToString(", "))
      }
    }

    return text.toString()
  }

  /**
   * Performs semantic search for operators based on task description.
   *
   * @param taskDescription the task description to search for
   * @return list of tool specs ordered by similarity
   */
  override fun searchByTask(taskDescription: String): List<ToolSpec> {
    val maxResults = 3

    return try {
      val queryEmbedding = embeddingService.generateEmbedding(taskDescription)
      val pageable = PageRequest.of(0, maxResults)

      val results = embeddingRepository.findTopByEmbeddingSimilarity(queryEmbedding, pageable)

      results.mapNotNull { embedding ->
        operatorsByName[embedding.operatorName]?.spec() // Single source of truth!
      }
    } catch (e: Exception) {
      log.error("Semantic search failed for query: {}", taskDescription, e)
      emptyList()
    }
  }

  override fun get(name: String): Operator? {
    return operatorsByName[name]
  }

  override fun size(): Int {
    return operatorsByName.size
  }

  // Package-private methods for testing
  internal fun getOperatorsByName(): Map<String, Operator> {
    return operatorsByName
  }

  internal fun createEmbeddingTextForTesting(spec: ToolSpec): String {
    return createEmbeddingText(spec)
  }
}